core/src/main/kotlin/com/strumenta/kolasu/ids/DeclarativeNodeIdProvider.kt

```diff
@@ -0,0 +1,78 @@
+package com.strumenta.kolasu.ids
+
+import com.strumenta.kolasu.model.Node
+import kotlin.reflect.KClass
+import kotlin.reflect.full.isSubclassOf
+import kotlin.reflect.full.isSuperclassOf
+
+/**
+ * Declarative node id provider instances allow to define
+ * language-specific or custom identifier construction policies.
+ *
+ * ```kotlin
+ * object MyDeclarativeNodeIdProvider: DeclarativeNodeIdProvider(
+ *     idFor(PackageDeclaration::class) { it.node.name }
+ *     idFor(ClassDeclaration::class) { (node: ClassDeclaration, identifiers: NodeIdProvider) ->
```

Comment:
Is the second parameter of the lambda (`identifies`) always corresponding the the instance of the NodeIdProvider being defined (`MyDeclarativeNodeIdProvider` in this case)?
If so, can we add it through `this`?

Reply:
Yes, we could move the node id provider as `this`, move `node` as `it` and remove the context class. Do you something like this?
```kotlin
NodeIdProvider.(NodeTy) -> String
```

Reply:
No, I was just thinking that:

```
object MyDeclarativeNodeIdProvider: DeclarativeNodeIdProvider(
      idFor(PackageDeclaration::class) { /* here I can see the NodeIdProvider using this, I do not need to receive it as parameter of the lambda */ }
}
```

Reply:
Exactly, this is now implemented :)